### PR TITLE
Merge link recovery based on reviews and pr feedack

### PR DESCRIPTION
**Pull request by vinaysurya**
This PR is created on top of my other PR (syncLinkRecoveryBranchWithMaster, which contains only changes necessary to sync with original master branch).

This PR has changes based on my reading of code and the interfaces we can restrict to. It also captures some of Xin's feedback when I discussed the changes with him. 

One of the primary changes being reducing two interfaces of ITerminusManager and IStore and combining them into one store. When looked at from a library User point of view it made sense to combine them and and allow folks to implement and pass in one interface than two.

Also, another change made is to combine the GetLinkTerminus to include getting the entire terminus details at one shot including the terminus's unsettled deliveries as well.

### CODE REVIEW

File: src/IAmqpTerminusStore.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        #endregion
+
+        #region DeliveryManagement
+
+        /// <summary>
+        /// Save a collection of <see cref="Delivery"/> under the given link terminus to be retrieved later.
+        /// </summary>
+        /// <param name="linkTerminus">The link terminus associated with the deliveries.</param>
+        /// <param name="unsettledDeliveries">A dictionary of Key-Value pairs of DeliveryTag and corresponding delivery.</param>
+        /// <returns>True if the deliveries were successfully saved under the link terminus.</returns>
```
Comment (bainian12345) [inline src/IAmqpTerminusStore.cs]: No longer returning true

File: src/IAmqpTerminusStore.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        /// <param name="linkTerminus">The link terminus object to be removed.</param>
+        /// <returns>True if the linkTerminus object was successfully found and removed, false otherwise.</returns>
+        Task<bool> TryRemoveLinkTerminusAsync(AmqpLinkIdentifier linkIdentifier, AmqpLinkTerminus linkTerminus);
+
+        /// <summary>
+        /// Try to get the link terminus with the given link identifier. Return true if the link terminus has been found.
+        /// </summary>
+        /// <param name="linkIdentifier">The unique identifier of a link endpoint which will be used to identify the link terminus.</param>
+        /// <param name="linkTerminus">The link terminus object associated with the link identifier.</param>
+        /// <returns>True if there is a link terminus object was found associated with the given link identifier.</returns>
```
Comment (bainian12345) [inline src/IAmqpTerminusStore.cs]: nit: grammar

File: src/AmqpLinkTerminus.cs [hunk 3]
```diff
@@ -203,15 +194,14 @@ internal protected bool TryAssociateLink(AmqpLink link, out AmqpLink stolenLink)
...
-            if (retrievedUnsettledDeliveries != null)
+            if (this.UnsettledDeliveries != null)
             {
                 if (link.Settings.Unsettled == null)
                 {
                     link.Settings.Unsettled = new AmqpMap(new Dictionary<ArraySegment<byte>, Delivery>(), MapKeyByteArrayComparer.Instance);
                 }
 
-                foreach (var kvp in retrievedUnsettledDeliveries)
+                foreach (var kvp in UnsettledDeliveries)
```
Comment (bainian12345) [inline src/AmqpLinkTerminus.cs]: nit: this.UnsettledDeliveries

File: src/AmqpLinkTerminus.cs [hunk 4]
```diff
@@ -143,14 +135,13 @@ public void Dispose()
...
-            IDictionary<ArraySegment<byte>, Delivery> localUnsettledDeliveries = await this.UnsettledDeliveryStore.RetrieveDeliveriesAsync(this);
             if (this.Identifier.IsReceiver)
             {
-                return localUnsettledDeliveries;
+                return this.UnsettledDeliveries;
             }
             else
             {
-                return await NegotiateUnsettledDeliveriesFromSender(remoteAttach, localUnsettledDeliveries);
+                return await NegotiateUnsettledDeliveriesFromSender(remoteAttach, this.UnsettledDeliveries);
```
Comment (bainian12345) [inline src/AmqpLinkTerminus.cs]: nit: this.NegotiateUnsettledDeliveriesFromSender

File: src/AmqpInMemoryTerminusStore.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// This class is the in-memory implemenation of <see cref="IAmqpTerminusStore"/>
+    /// Will be used to store and retrieve both linkTerminus Information and deliveries associated with the termini.
+    /// </summary>
+    public class AmqpInMemoryTerminusStore : IAmqpTerminusStore
+    {
+        readonly object thisLock;
```
Comment (bainian12345) [inline src/AmqpInMemoryTerminusStore.cs:16]: This lock seems to be pretty fat, I'm not sure if we can separate out the locking between the terminus operations and the deliveries operations for a terminus?

**Review by bainian12345 — COMMENTED** [review]
Looks good, but I don't know if the current locking mechanism will be too impactful for performance.

File: src/IAmqpTerminusStore.cs (hunk 1 above)
Comment (vinaysurya) [reply src/IAmqpTerminusStore.cs]: Fixed to return

File: src/AmqpInMemoryTerminusStore.cs (hunk 5 above)
Comment (vinaysurya) [reply src/AmqpInMemoryTerminusStore.cs:16]: Yeah, it should certainly be possible Brian. Infact, we should be able to perform a lot of optimizations with this implementation of the store.
1. Make the terminus lock per containerid.
2. Separate the senders and receiver link terminus are different, we could maintain those as separate collections in the store.
3. Separate the lock for terminus management versus delivery management operations.

Will do the optimizations over time. For now I want to get a version out there, so SDK team can develop in parallel.

Also, mostly the concern is about the RemoveUnsettledDelivery, which can be called on every transfer. Other than most the rest of the code is only called during link attach, etc so its not in the performance path.  Also most locks are in memory should not be too much concern for the immediate future. Will revisit this once we get a PR out in the main branch, so we can get the reviews and the SDK work going in parallel.

**Pull request merged**